src/ucs/datastruct/mpmc.c

```diff
@@ -17,82 +17,66 @@
 #include <ucs/debug/memtrack_int.h>
 
 
-ucs_status_t ucs_mpmc_queue_init(ucs_mpmc_queue_t *mpmc, uint32_t length)
+ucs_status_t ucs_mpmc_queue_init(ucs_mpmc_queue_t *mpmc)
 {
-    uint32_t i;
+    ucs_status_t ret;
 
-    mpmc->length   = ucs_roundup_pow2(length);
-    mpmc->shift    = ucs_ilog2(mpmc->length);
-    if (mpmc->shift >= UCS_MPMC_VALID_SHIFT) {
-        return UCS_ERR_INVALID_PARAM;
+    ret = ucs_spinlock_init(&mpmc->lock, 0);
+    if (ret == UCS_OK) {
+        ucs_queue_head_init(&mpmc->queue);
     }
 
-    mpmc->consumer = 0;
-    mpmc->producer = 0;
-    mpmc->queue = ucs_malloc(sizeof(*mpmc->queue) * mpmc->length, "mpmc");
-    if (mpmc->queue == NULL) {
-        return UCS_ERR_NO_MEMORY;
-    }
-
-    for (i = 0; i < mpmc->length; ++i) {
-        mpmc->queue[i] = UCS_BIT(UCS_MPMC_VALID_SHIFT);
-    }
-
-    return UCS_OK;
+    return ret;
 }
 
 void ucs_mpmc_queue_cleanup(ucs_mpmc_queue_t *mpmc)
 {
-    ucs_free(mpmc->queue);
-}
+    ucs_queue_elem_t *elem;
 
-static inline uint64_t __ucs_mpmc_queue_valid_bit(ucs_mpmc_queue_t *mpmc, uint32_t location)
-{
-    return ((uint64_t)location >> mpmc->shift) & 1;
+    while ((elem = ucs_queue_pull(&mpmc->queue))) {
+        ucs_free(ucs_container_of(elem, ucs_mpmc_elem_t, super));
+    }
 }
 
 ucs_status_t ucs_mpmc_queue_push(ucs_mpmc_queue_t *mpmc, uint64_t value)
 {
-    uint32_t location;
+    ucs_mpmc_elem_t *elem;
+
+    elem = ucs_malloc(sizeof(ucs_mpmc_elem_t), "mpmc elem");
+    if (elem == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
 
-    ucs_assert((value >> UCS_MPMC_VALID_SHIFT) == 0);
+    elem->value      = value;
+    elem->super.next = NULL;
```

Comment:
not needed; the linked list is handled by ucs_queue functions

Reply:
done